hazelcast/src/test/java/com/hazelcast/executor/CompletableFutureTest.java

```diff
@@ -23,158 +23,350 @@
-        startLatch = new CountDownLatch(1);
-        ref1 = new AtomicReference<Object>();
-        ref2 = new AtomicReference<Object>();
-    }
-
-    @After
-    public void tearDown() {
+        startLogicLatch = new CountDownLatch(1);
+        executedLogic = new CountDownLatch(1);
+        finishedLatch = new CountDownLatch(1);
+        inExecutionLatch = new CountDownLatch(1);
+        reference1 = new AtomicReference<Object>();
+        reference2 = new AtomicReference<Object>();
     }
 
     @Test
     public void preregisterCallback() throws Exception {
-        doneLatch = new CountDownLatch(1);
-        final ICompletableFuture<String> f = submit(awaitStartLatch());
-        f.andThen(setRefAndBumpDoneLatch(ref1));
-        startLatch.countDown();
-        assertOpenEventually(doneLatch);
-        assertEquals("success", ref1.get());
+        ICompletableFuture<String> f = submitAwaitingTask(expectedNumberOfCallbacks(1), NO_EXCEPTION);
+        f.andThen(storeTaskResponseToReference(reference1));
+
+        releaseAwaitingTask();
+
+        assertCallbacksExecutedEventually();
+        assertEquals("success", reference1.get());
     }
 
     @Test
     public void preregisterTwoCallbacks() throws Exception {
-        doneLatch = new CountDownLatch(2);
-        final ICompletableFuture<String> f = submit(awaitStartLatch());
-        f.andThen(setRefAndBumpDoneLatch(ref1));
-        f.andThen(setRefAndBumpDoneLatch(ref2));
-        startLatch.countDown();
-        assertOpenEventually(doneLatch);
-        assertEquals("success", ref1.get());
-        assertEquals("success", ref2.get());
+        ICompletableFuture<String> f = submitAwaitingTask(expectedNumberOfCallbacks(2), NO_EXCEPTION);
+        f.andThen(storeTaskResponseToReference(reference1));
+        f.andThen(storeTaskResponseToReference(reference2));
+
+        releaseAwaitingTask();
+
+        assertCallbacksExecutedEventually();
+        assertEquals("success", reference1.get());
+        assertEquals("success", reference2.get());
     }
 
     @Test
-    public void preregisterTwoCallbacks_withFailure() throws Exception {
-        doneLatch = new CountDownLatch(2);
-        final ICompletableFuture<String> f = submit(awaitStartLatch(), throwException());
-        f.andThen(setRefAndBumpDoneLatch(ref1));
-        f.andThen(setRefAndBumpDoneLatch(ref2));
-        startLatch.countDown();
-        assertOpenEventually(doneLatch);
-        assertTestException(ref1, ref2);
+    public void preregisterTwoCallbacks_taskThrowsException() throws Exception {
+        ICompletableFuture<String> f = submitAwaitingTask(expectedNumberOfCallbacks(2), THROW_TEST_EXCEPTION);
+        f.andThen(storeTaskResponseToReference(reference1));
+        f.andThen(storeTaskResponseToReference(reference2));
+
+        releaseAwaitingTask();
+
+        assertCallbacksExecutedEventually();
+        assertTestExceptionThrown(reference1, reference2);
     }
 
     @Test
+    @Repeat(10)
```

Comment:
Thx for spotting - I left it intentionally there just increase to chance of a failing test should there be any hidden bug. Would you prefer to have all the `@Repeat`s deleted? These tests are rather fast, so it's not a huge bottleneck